### PR TITLE
rpc: Add experimental config params to allow for subscription buffer size control (tm v0.34.x)

### PR DESCRIPTION
**Pull request by thanethomson**
Follows from #7188 and related to #6729.

The aim here is to provide a short-term workaround for #6729 in the v0.34.x series of Tendermint that allows for tuning of the internal buffer sizes relating to event subscription. The following configuration parameters are introduced in the `[rpc]` section:

```toml
[rpc]

# Experimental parameter to specify the maximum number of events a node will
# buffer, per subscription, before returning an error and closing the
# subscription. Must be set to at least 100, but higher values will accommodate
# higher event throughput rates (and will use more memory).
experimental_subscription_buffer_size = 1000

# Experimental parameter to specify the maximum number of RPC responses that
# can be buffered per WebSocket client. If clients cannot read from the
# WebSocket endpoint fast enough, they will be disconnected, so increasing this
# parameter may reduce the chances of them being disconnected (but will cause
# the node to use more memory).
#
# Must be at least the same as "experimental_subscription_buffer_size",
# otherwise connections could be dropped unnecessarily. This value should
# ideally be somewhat higher than "experimental_subscription_buffer_size" to
# accommodate non-subscription-related RPC responses.
experimental_websocket_write_buffer_size = 1000

# If a WebSocket client cannot read fast enough, at present we may
# silently drop events instead of generating an error or disconnecting the
# client.
#
# Enabling this experimental parameter will cause the WebSocket connection to
# be closed instead if it cannot read fast enough, allowing for greater
# predictability in subscription behaviour.
experimental_close_on_slow_client = false
```

### CODE REVIEW

**Comment by thanethomson** [comment]
Seems like the new linting changes in https://github.com/tendermint/tendermint/commit/8a2dcbafaec0eddd960af4c7fe804a8044999c22 are picking up a whole bunch of spelling mistakes throughout the codebase :slightly_smiling_face:

**Comment by tychoish** [comment]
> Seems like the new linting changes in 8a2dcba are picking up a whole bunch of spelling mistakes throughout the codebase ￼

They're less mistakes and more conventions :p 

Anyway, those should be fixed if you rebase/etc.

**Comment by thanethomson** [comment]
So we've now managed to confirm that:
1. With the latest version of Hermes, we can still reproduce the RPC disconnection with the official Gaia v4.2.1 release (Tendermint v0.34.9).
2. We can reproduce the RPC disconnection with Gaia v4.2.1 built with Tendermint from the `v0.34.x` branch.
3. We can reproduce the RPC disconnection with Gaia v4.2.1 built with Tendermint from this PR's branch (`thane/6729-websocket-bufsize2`) with experimental buffer sizes set to 100.
4. Using the same setup as in (3), but setting the buffer sizes to 1000 results in us not being able to reproduce the RPC disconnection. We do not get disconnected from Gaia, even when submitting 10,000 transactions as quickly as we can with Hermes (all processes running on the same machine).

Beyond enabling these experimental config parameters, I'd like to propose we increase the default subscription buffer size to 1000 to better handle the loads we're seeing in live networks.

File: rpc/core/events.go [hunk 1]
```diff
@@ -37,11 +33,13 @@ func Subscribe(ctx *rpctypes.Context, query string) (*ctypes.ResultSubscribe, er
 	subCtx, cancel := context.WithTimeout(ctx.Context(), SubscribeTimeout)
 	defer cancel()
 
-	sub, err := env.EventBus.Subscribe(subCtx, addr, q, subBufferSize)
+	sub, err := env.EventBus.Subscribe(subCtx, addr, q, env.Config.SubscriptionBufferSize)
 	if err != nil {
 		return nil, err
 	}
 
+	closeIfSlow := env.Config.CloseOnSlowClient
```
Comment (creachadair) [inline rpc/core/events.go:41]: Is this change necessary? I see the point, but it changes the semantics of the client in subtle ways. Did your experiments require this to be set to mitigate the disconnects? If not, could we reasonably defer this feature for now?

**Review by creachadair — COMMENTED** [review]
Overall this change seems mechanically fine. My one main concern is this: If we add these experimental settings _only_ to the v0.34 line, it becomes an upgrade blocker for v0.35.

The biggest users are the ones most likely to need this knob, and there are enough other quality of life improvements in the v0.35 series that we'd prefer not to add any additional impediments to their upgrading.

It looks to me like the config plumbing parts of this PR could be safely added to v0.35.x for the next point release, which would avert that. I think it'd be fine to land them separately rather than as a backport from v0.35, though.

Your thoughts?

**Comment by creachadair** [comment]
> Overall this change seems mechanically fine. My one main concern is this: If we add these experimental settings _only_ to the v0.34 line, it becomes an upgrade blocker for v0.35.
> 
> The biggest users are the ones most likely to need this knob, and there are enough other quality of life improvements in the v0.35 series that we'd prefer not to add any additional impediments to their upgrading.
> 
> It looks to me like the config plumbing parts of this PR could be safely added to v0.35.x for the next point release, which would avert that. I think it'd be fine to land them separately rather than as a backport from v0.35, though.
> 
> Your thoughts?

Just to clarify: I'm not objecting to merging this change into v0.34 either way, I am just thinking about the path for someone who needs to upgrade.

File: config/config.go [hunk 2]
```diff
@@ -52,6 +52,9 @@ var (
 
 	defaultNodeKeyPath  = filepath.Join(defaultConfigDir, defaultNodeKeyName)
 	defaultAddrBookPath = filepath.Join(defaultConfigDir, defaultAddrBookName)
+
+	minSubscriptionBufferSize     = 100
+	defaultSubscriptionBufferSize = 1000
```
[the remark this answers is not in the document]
Comment (thanethomson) [reply config/config.go:57]: Sure, we can drop this to a more modest default value. How many subscribers do people usually see in a full node or validator though?

File: rpc/core/events.go (hunk 1 above)
Comment (thanethomson) [reply rpc/core/events.go:41]: The default behaviour here doesn't change at all (apart from the evaluation of the newly introduced `if` statement).

This change shows what I imagine the code should look like if we're really going to be closing the connection on slow clients. Until recently, when we inspected the code here, we didn't know that the node could just silently drop responses - something that could result in all kinds of nondeterministic behaviour for IBC.

So I'd say yes, this change is definitely necessary for us.

**Comment by thanethomson** [comment]
> It looks to me like the config plumbing parts of this PR could be safely added to v0.35.x for the next point release, which would avert that. I think it'd be fine to land them separately rather than as a backport from v0.35, though.

I'd be more than happy to submit a separate PR to introduce similar changes into v0.35.

**Comment by thanethomson** [comment]
Are the coverage tests known to be non-deterministic? It seems like each time I re-run the tests, a different set of coverage tests fail.

**Comment by creachadair** [comment]
> Are the coverage tests known to be non-deterministic? It seems like each time I re-run the tests, a different set of coverage tests fail.

Some of the end-to-end tests are nondeterministic, but the unit tests should usually not be. This looks like a bug in the tests, though it is quite possible it's a bug we fixed already on trunk. Over the past few months we've cleaned up various issues with goroutine lifetime.

If it doesn't fail consistently, I don't think we should block over it.

File: config/config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (creachadair) [reply config/config.go]: > Sure, we can drop this to a more modest default value. How many subscribers do people usually see in a full node or validator though?

I don't know—I suspect it depends a lot on the size of the network and the behaviour of the application. I mainly want to avoid blowing up nodes that didn't already need this new knob.

**Review by creachadair — APPROVED** [review]
